src/main/java/com/shapesecurity/shift/es2016/template/Template.java

```diff
@@ -0,0 +1,207 @@
+import com.shapesecurity.shift.es2016.parser.SourceSpan;
+import com.shapesecurity.shift.es2016.reducer.Director;
+import com.shapesecurity.shift.es2016.reducer.Flattener;
+import com.shapesecurity.shift.es2016.reducer.ReconstructingReducer;
+import com.shapesecurity.shift.es2016.reducer.Reducer;
+import com.shapesecurity.shift.es2016.reducer.WrappedReducer;
+import com.shapesecurity.shift.es2016.utils.WithLocation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Template {
+	public static class NodeInfo {
+		public final String name;
+		public final Node node;
+		public final ParserWithLocation.Comment comment;
+
+		NodeInfo(String name, Node node, ParserWithLocation.Comment comment) {
+			this.name = name;
+			this.node = node;
+			this.comment = comment;
+		}
+	}
+
+	static class Marker {
+		final String name;
+		final Maybe<Class<? extends Node>> type;
+		final int start;
+		final ParserWithLocation.Comment comment;
+
+		Marker(String name, Maybe<Class<? extends Node>> type, int start, ParserWithLocation.Comment comment) {
+			this.name = name;
+			this.type = type;
+			this.start = start;
+			this.comment = comment;
+		}
+	}
+
+	public static Program replace(Program tree, F2<Node, Node, Maybe<Node>> getReplacement) {
+		Reducer<Node> replacer = new WrappedReducer<>(
+			(originalNode, newNode) -> getReplacement.apply(newNode, originalNode).orJust(newNode),
+			new ReconstructingReducer() // TODO LazyCloneReducer
+		);
+		return (Program) Director.reduceProgram(replacer, tree);
+	}
+
+	static final Pattern commentRegex = Pattern.compile("^# ([^#]+) (?:# ([^#]+) )?#$");
+	public static Maybe<Pair<String, Maybe<Class<? extends Node>>>> parseComment(String text) {
+		Matcher matcher = commentRegex.matcher(text);
+		if (!matcher.matches()) {
+			return Maybe.empty();
+		}
+
+		String name = matcher.group(1);
+		String typeName = matcher.group(2);
+
+		if (typeName != null) {
+			try {
+				Class type = Class.forName("com.shapesecurity.shift.es2016.ast." + typeName); // This is kind of awful, but... it does work...
+				@SuppressWarnings("unchecked")
+				Class<? extends Node> nodeType = (Class<? extends Node>) type;
+				return Maybe.of(Pair.of(name, Maybe.of(nodeType)));
+			} catch (ClassNotFoundException | ClassCastException e) {
+				throw new IllegalArgumentException("Unrecognized node type \"" + typeName + "\"");
+			}
+		}
+
+		return Maybe.of(Pair.of(name, Maybe.empty()));
+	}
+
+	public static ImmutableList<NodeInfo> findNodes(Program tree, WithLocation locations, ImmutableList<ParserWithLocation.Comment> comments) {
+		// TODO: allow custom matcher
```

Comment:
Let's do this TODO or at least change our API so that we won't need to break the API later when we do this TODO.

Reply:
I was planning on just adding an overload which this function would defer to, but sure, easy enough to add it now given how things ended up being factored.